### ISSUE TITLE
The new code editor rules are too opinionated

### ISSUE THREAD

**Issue by alphalambda** [google/codeworld#1280]
**Shared Program**: https://code.world/#Pxxxxxxxxxxx (if applicable)

https://code.world/#Pr6Al3aiA0w4JLvCE73GqLg

**Description**

As the example indicates, the new editor rules make it very difficult to have lists properly aligned, no matter whether the comma is at the beginning or at the end. This is just an example of many formatting rules that are too opinionated now.

I understand the intention of enforcing a consistent style, but the style currently enforced is too opinionated and puts a high burden on teacher trainers, because it does not match the style rules I taught when I trained the teachers. Now the teachers need to be either retrained (which has a high cost) or let to fight with an editor that does not want to format things the way they were taught.

There should be a way to disable rules that enforce a particular style that may not coincide with the style chosen by other curriculum providers.

(If you don't know what's causing the problem, please include the
browser and operating system you're using in the description.)

**Comment by cdsmith** [comment]
I'm quite willing to change the indent rules.   I don't want to disable them, though, because they actually prevented a LOT of student errors the last time I taught with them.  Whole classes of errors just disappeared.

The one tricky constraint is that indentation of a line can depend on what's before it, but not what's after.  That means, for example, that in a case like:

    program = drawingOf(translated(colored(circle(5),
                    blue), 3, 4))

The indentation of the second line must be decided before the contents of the line are known (exception: the very *first* token of the line can affect indentation).  This explains why it's indented three levels there, because you might have written this instead:

    program = drawingOf(translated(colored(circle(5),
                    blue)
                , 3, 4),
            foo
        )
        & coordinatePlane

(Yes, I know `drawingOf` doesn't take two arguments, but the editor doesn't look up types when it indents, so as far as it knows, it might.)

So far, layout follows a simple policy: one indent level per context (either a layout context, or a set of explicit brackets), and an extra indent level for continuing the same line.  Certain continuations are marked as half-indents (e.g., `where` clauses), because they just looked better to me that way.  I'm quite happy to further tweak the rules so that they are more acceptable.

You can, of course, disable the auto-indents in your server.  Let me know if you want help doing so.  As I mentioned above, I think you'll find they help quite a lot with new students.  Let's see if I can make the auto-indent more agreeable to you.

**Comment by cdsmith** [comment]
Do you have the rules you've taught to your teacher trainers?  I'll see if I can just implement those.

**Comment by alphalambda** [comment]
I am interested in testing whether I see a similar reduction in student errors, but I think the current implementation is too eager, to the point of frankly being frustrating for me. I am currently the only one using the new editor in my server, but I need to make a decision on whether to deploy it or not before the Spring semester begins.

The most frustrating points are:

1. The backspace key should erase the character immediately before it, just like the space bar inserts one space. The backspace now shifts the line to the recommended position, ignoring my intention to override this. If I wanted to shift the line to the previous TAB stop, I would use shift-Tab.
2 After I modify the indentation manually, the editor should not override my decision and re-indent it. This is especially annoying when I am trying to add comments starting at a certain column, and every time I type a new character, the line moves on me!
3 Indentation for lists is wrong, as my example above shows. I prefer commas at the beginning, but no matter where they are, the list items should be aligned to each other. I insisted on this a lot during the training.

**Comment by cdsmith** [comment]
> The backspace key should erase the character immediately before it.

Backspace actually doesn't re-indent.  It decreases the indent to the previous level.  I can make backspace less aggressive, though.  I didn't actually realize it was triggering when the cursor is before the indent point.

I think this should be the intended behavior:

1. If the current cursor is at the first non-whitespace character of a line, *and* the current indent is a standard one (meaning it's one you can reach with auto-indent and tabs), then unindent one level.
2. In any other case, act like backspace.

I do want to leave case 1.  I know that you can use Shift-Tab, but students will not.  Let's see if I can make it non-annoying without losing that.

> After I modify the indentation manually, the editor should not override my decision and re-indent it.

This one is a great idea, but tricky because it means there's extra document state, and CodeMirror doesn't make it easy to keep document state on the changing document, unless it's computed from tokens.  This isn't.

An easier rule to implement would be: don't do electric indent (that is the indents that happen when you're not starting a new line) unless the line was already indented at the "correct" level according to auto-indent.  I'll start with that.

One case that won't fix:

    complete line of code
        -- comment

When typing the second line, the first dash will (after this change) not trigger auto-indent because the indentation was non-standard before typing it.   But the second dash would trigger auto-indent still, because the document is identical to the one you would have obtained by typing the dash in the first column.  So you would lose the indent.

Another alternative is to keep a disable-auto-indent flag, but only for the current line.  If you move to a new line and back, it would be reenabled.

> Indentation for lists is wrong, as my example above shows. I prefer commas at the beginning, but no matter where they are, the list items should be aligned to each other.

I love the idea of aligning lists and tuples with the first character inside them, and it should be easy to implement, even sharing much of the machinery from layout.  As far as I know, auto-indent has no preference for whether you place commas at the beginning or end of lines, unless you want a rule (which would also be easy) to un-indent the commas by two spaces so the elements will align?

**Comment by alphalambda** [comment]
On Mon, Dec 30, 2019 at 06:52:46AM -0800, Chris Smith wrote:

> I can make backspace less aggressive

Yes, please!

> I do want to leave case 1.  Let's see if I can make it non-annoying without losing that.

Thank you! I'm not sure whether it is possible to meet those two goals at the same time, but I hope not-too-annoying would be
good enough.


> Another alternative is to keep a disable-auto-indent flag, but only for the current line.
> If you move to a new line and back, it would be reenabled.

This sounds like a potentially satisfactory solution.


> I love the idea of aligning lists and tuples with the first character inside them, and it should be easy to implement

Great!

About placement of commas, I prefer them at the beginning, but I am alone on that. Everyone else dislikes that choice.
So, my advice so far has been "place the commas wherever it feels OK for you, but nevertheless align the items properly"

**Comment by cdsmith** [comment]
Status update:

I have pushed changes to make backspace less aggressive about indent points, and to align elements in all brackets.  It took a lot of tweaking, but I think it works reasonably now.

The one feature request that's still outstanding here is avoiding electric indents when the indent has been changed.  I'm still planning on doing that, but it may be a bit trickier than I anticipated.

**Comment by cdsmith** [comment]
@alphalambda I've implemented all of these mitigations, so this is closed.  Feel free to reopen if you still have trouble.

Note: #1281 and #1284 are known issues I discovered while working on this.  I'm still planning to fix them, but it's more useful to me to have issues for specific action items when there's something definite to be done.

**Comment by alphalambda** [comment]
Thanks! I'll try it again.

Happy New Year!

**Comment by cdsmith** [comment]
Thanks for pushing back on this.  I'm glad I worked on it, and I think this is in a much better place than it was last semester.  I'm pretty excited by it, and also looking forward to trying a rule that students must press Ctrl-I (which now reindents the code) before asking for help on a syntax error.
